File: Python/Dysk.py
def encode_to_disk(data):
    disk = ['D', 'D', 'D', 'D']
    for bit in data:
        prev = disk[-1]
        if bit == '1':
            disk.append('G' if prev == 'D' else 'D')
        else:
            disk.append(prev)
    return ''.join(disk)

def decode_from_disk(disk):
    data = []
    for i in range(4, len(disk)):
        prev = disk[i - 1]
        current = disk[i]
        if prev != current:
            data.append('1')
        else:
            data.append('0')
    return ''.join(data)

data = input()
if all(c in '01' for c in data):
    print(encode_to_disk(data))
else:
    print(decode_from_disk(data))

File: Java/Dysk.java
import java.util.Scanner;

public class Main
{
    public static String encodeToDisk(String data) 
    {
        StringBuilder disk = new StringBuilder("DDDD");
        for (char bit : data.toCharArray()) {
            char prev = disk.charAt(disk.length() - 1);
            if (bit == '1')
                disk.append(prev == 'D' ? 'G' : 'D');
            else
                disk.append(prev);
        }
        return disk.toString();
    }

    public static String decodeFromDisk(String disk) 
    {
        StringBuilder data = new StringBuilder();
        for (int i = 4; i < disk.length(); i++) 
        {
            char prev = disk.charAt(i - 1);
            char current = disk.charAt(i);
            if (prev != current)
                data.append('1');
            else
                data.append('0');
        }
        return data.toString();
    }

    public static void main(String[] args) 
    {
        Scanner input = new Scanner(System.in);
        String data = input.nextLine();
        if (data.matches("[01]+"))
            System.out.println(encodeToDisk(data));
        else if (data.matches("[GD]+"))
            System.out.println(decodeFromDisk(data));
        input.close();
    }
}